src/main/java/dev/morling/onebrc/CalculateAverage_rby.java

```diff
@@ -0,0 +1,223 @@
+            while(iter.hasNext()) {
+                var e = iter.next();
+                var ix = e.getValue().intValue() * 4;
+                var avg = Math.round(stats[ix + 2]/((double)stats[ix+3]))/10.0;
+                System.out.print(", " + e.getKey() + "="
+                        + (stats[ix]/10.0) + "/"
+                        + avg + "/"
+                        + (stats[ix + 1]/10.0)) ;
+            }
+            System.out.println("}");
+        }
+    }
+
+    static final int MAX_CITIES = 1000;
+    static final int ARRAY_SIZE = 1 << 20;
+
+    static Stats stats(Path p, long start, long end) {
+        int nextCityIx = 0;
+        var cityIndexes = new HashMap<String, Integer>(MAX_CITIES, 1.0f);
+        int[] stats = new int[MAX_CITIES * 4];
+        for (int i = 0; i < MAX_CITIES; i++) {
+            stats[i * 4] = Integer.MAX_VALUE;
+            stats[i * 4 + 1] = Integer.MIN_VALUE;
+        }
+
+        try {
+            final var channel = (FileChannel) Files.newByteChannel(p, EnumSet.of(StandardOpenOption.READ));
+            channel.position(start);
+            var offset = start;
+            final byte[] array = new byte[ARRAY_SIZE];
+            // the next expected char, the most simple stateMachine
+            char nextChar = ';';
+            // good enough for a city name, or a double
+            byte[] strbuff = new byte[128];
+            int strbuffIx = 0;
+            int cityIndex = 0;
+            final var buffer = ByteBuffer.allocateDirect(CHUNK_SIZE);
+
+            while (offset < end) {
+                final int limit = channel.read(buffer);
+                if (limit <= 0)
+                    break;
+                offset += limit;
+                int totalRead = 0;
+                buffer.flip();
+                while (totalRead < limit) {
+                    int read = Math.min(array.length, limit - totalRead);
+                    buffer.get(array, 0, read);
+                    totalRead += read;
+
+                    for (int i = 0; i < read; i++) {
+                        if (nextChar == '\n' && array[i] == '.')
+                            continue;
+                        strbuff[strbuffIx++] = array[i];
+                        if (array[i] == nextChar) {
+                            var str = new String(strbuff, 0, strbuffIx - 1, "utf8");
+                            strbuffIx = 0;
+                            switch (nextChar) {
+                                case ';':
+                                    nextChar = '\n';
+                                    var mbCityIx = cityIndexes.get(str);
+                                    if (mbCityIx == null) {
+                                        cityIndex = nextCityIx;
+                                        cityIndexes.put(str, nextCityIx++);
+                                        if (nextCityIx * 4 >= stats.length) {
+                                            var newStats = Arrays.copyOf(stats, stats.length * 2);
```

Comment:
@gunnarmorling This is where more than 1000 cities is handled.
the array is expanded and copied.
In the main case of 1b with real cities names the number was less than 1000